lib/hanami/cli/commands/app/db/command.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+require "shellwords"
+require_relative "utils/database"
+
+module Hanami
+  module CLI
+    module Commands
+      module App
+        module DB
+          # Base class for `hanami` CLI commands intended to be executed within an existing Hanami
+          # app.
+          #
+          # @since 2.2.0
+          # @api private
+          class Command < App::Command
+            option :app, required: false, type: :boolean, default: false, desc: "Use app database"
+            option :slice, required: false, desc: "Use database for slice"
+
+            private
+
+            def databases(app: false, slice: nil)
+              if app
+                [database_for_app]
+              elsif slice
+                [database_for_slice(slice)]
+              else
+                all_databases
+              end
+            end
+
+            def database_for_app
+              Utils::Database[app]
+            end
+
+            def database_for_slice(slice)
+              slice = inflector.underscore(Shellwords.shellescape(slice)).to_sym
+
+              Utils::Database[app.slices[slice]]
+            end
+
+            def all_databases
+              slices = app.slices.with_nested << app
+
+              slices_by_database_url = slices.each_with_object({}) { |slice, hsh|
+                next unless slice.container.providers.find_and_load_provider(:db)
+
+                slice.prepare :db
+                database_url = slice["db.gateway"].connection.uri
+
+                hsh[database_url] ||= []
+                hsh[database_url] << slice
+              }
+
+              databases = slices_by_database_url.each_with_object([]) { |(url, slices), arr|
+                slices_with_config = slices.select { _1.root.join("config", "db").directory? }
+
+                database = Utils::Database[slices_with_config.first || slices.first]
+
+                warn_on_misconfigured_database database, slices_with_config
+
+                arr << database
+              }
+
+              databases
+            end
```

Comment:
This is a pretty important method. We'll likely want something like this for all our `db` subcommands, which is why I've put it here.

As we begin to use this from the other subcommands, I expect we may want to tweak it a little. We'll see.

This is also why I didn't try and factor this out into smaller methods. Each time I tried to do so, it felt like it impaired the ease of understanding exactly what's going on to prepare this list of all databases, so I've left it as a single large-ish method for now.